module/VuFind/src/VuFind/Controller/OverdriveController.php

```diff
@@ -258,41 +333,114 @@ public function holdAction()
+            $holdEmail = $hold->emailAddress;
         } elseif ($action == 'returnTitleConfirm') {
+            if (current($formats)->id == 'magazine-overdrive') {
+                $isMagazine = true;
+                //we don't have the cover of the actual issue here,
+                //but it's not worth going back to the API for it.
+                //just use current issue cover.
+            }
             $actionTitleCode = 'od_early_return';
-        } elseif ($action == 'getTitleConfirm') {
-            // Get only formats that are available...
-            $formats = $driver->getAvailableDigitalFormats();
-            $actionTitleCode = 'od_get_title';
+
+            // ACTION SECTION
         } elseif ($action == 'doCheckout') {
             $actionTitleCode = 'od_checkout';
+            if ($edition) {
+                $od_id = $edition;
+                $this->debug("checking out edition: $edition");
+            }
             $result = $this->connector->doOverdriveCheckout($od_id);
         } elseif ($action == 'placeHold') {
             $actionTitleCode = 'od_hold';
             $email = $this->params()->fromPost('email');
             $result = $this->connector->placeOverDriveHold($od_id, $email);
+            if ($result->status) {
+                $result->code = 'od_hold_place_success';
+                $result->codeParams = ['%%holdListPosition%%' => $result->data->holdListPosition];
+            } else {
+                $result->code = 'od_hold_place_failure';
+            }
+        } elseif ($action == 'editHoldEmail') {
+            $actionTitleCode = 'od_hold';
+            $email = $this->params()->fromPost('email');
+            $result = $this->connector->updateOverDriveHold($od_id, $email);
+
+            if ($result->status) {
+                $result->code = 'od_hold_update_success';
+            } else {
+                $result->code = 'od_hold_update_failure';
+            }
+        } elseif ($action == 'suspendHold') {
+            $actionTitleCode = 'od_susp_hold';
+            $suspendValue = $this->params()->fromPost('suspendValue');
+            $hold = $this->connector->getHold($od_id, false);
+            $holdEmail = $hold->emailAddress;
+            $suspensionType = $suspendValue == -1 ? 'indefinite' : 'limited';
+            $result = $this->connector->suspendHold($od_id, $holdEmail, $suspensionType, $suspendValue);
+
+            if ($result->status) {
+                if ($suspensionType == 'indefinite') {
+                    $result->code = 'od_hold_susp_indef';
+                } else {
+                    $result->code = 'od_hold_redelivery';
+                    $result->codeParams = ['%%days%%' => $result->data->holdSuspension->numberOfDays];
+                }
+            } else {
+                $result->code = 'od_hold_update_failure';
+            }
+        } elseif ($action == 'editSuspendedHold') {
+            $actionTitleCode = 'od_susp_hold_edit';
```

Comment:
This method is very long; it might be easier to read, troubleshoot and extend if it could be broken down into smaller parts. I certainly don't insist that you do this right now, but it might be helpful. For example, might it be possible to break down this long if/else statement into something like this?

```
$actions = [
    'editSuspendedHold' => ['titleCode' => 'od_susp_hold_edit', 'resultsMethod' => 'getEditSuspendedCodeResult'],
    // other action definitions go here...
];
if (isset($actions[$action])) {
    $actionTitleCode = $actions[$action]['titleCode'];
    $result = ($this->{$actions[$action]['resultsMethod']})();
} else {
    // action not defined logic goes here
}
```

And then you end up with a whole lot of small action processing methods instead of one big one.

Maybe there are reasons it's not as simple as this, but just an idea. If you like the idea but don't have time to deal with this right now, I'm happy to get this merged as-is and can help with refactoring in the future, as long as you can help with testing. :-)

Reply:
I like the idea. I've been meaning to but just haven't gotten up the gumption yet. I'll start working on it.